flatdata-cpp/src/TarReader.cpp

```diff
@@ -0,0 +1,192 @@
+/**
+ * Copyright (c) 2021 HERE Europe B.V.
+ * See the LICENSE file in the root of this project for license details.
+ */
+
+#include <flatdata/internal/TarReader.h>
+
+#include <array>
+#include <cassert>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <stdexcept>
+
+namespace flatdata
+{
+namespace internal
+{
+namespace
+{
+/* tar Header Block, from POSIX 1003.1-1990.  */
+struct TarFileHeader
+{                         /* byte offset */
+    char name[ 100 ];     /*   0 */
+    char mode[ 8 ];       /* 100 */
+    char uid[ 8 ];        /* 108 */
+    char gid[ 8 ];        /* 116 */
+    char size[ 12 ];      /* 124 */
+    char mtime[ 12 ];     /* 136 */
+    char chksum[ 8 ];     /* 148 */
+    char typeflag;        /* 156 */
+    char linkname[ 100 ]; /* 157 */
+    char magic[ 6 ];      /* 257 */
+    char version[ 2 ];    /* 263 */
+    char uname[ 32 ];     /* 265 */
+    char gname[ 32 ];     /* 297 */
+    char devmajor[ 8 ];   /* 329 */
+    char devminor[ 8 ];   /* 337 */
+    char prefix[ 155 ];   /* 345 */
+    char padding[ 12 ];   /* 500 */
+                          /* 512 */
+};
+
+/* Values used in typeflag field.  */
+const char REGTYPE = '0';           /* regular file */
+const char AREGTYPE = '\0';         /* regular file */
+const char DIRTYPE = '5';           /* directory */
+const char GNUTYPE_LONGNAME = 'L';  // Identifies the *next* file on the tape as having a long name.
+
+uint64_t
+decode_numeric_field( const char* data, size_t size )
+{
+    const uint64_t MAX_VALUE = std::numeric_limits< size_t >::max( );
+
+    if ( *data & 0x80 )  // Highest byte set indicates base-256 encoding
+    {
+        uint64_t value = ( *data & 0x7F );
+        for ( size_t idx = 1; idx < size; ++idx )
+        {
+            if ( value > MAX_VALUE / 256 )
+            {
+                throw std::runtime_error( "Numeric value too large" );
```

Comment:
This kind of error we should most likely catch in the `create` method: FileResourceStorage does not throw exceptions on opening (I think).

Also: Maybe some more context for the error message (error opening TAR file, etc)?

PPS: Flatdata is currently a bit bad on proper error feedback channels in C++ (Rust version is a bit better, but also a bit worse on some regards)

Reply:
See above: What is the recommended way to propagate error messages in flatdata?